File: YouboraLib/persistence/infinity/YBInfinityLocalManager.m
//
//  YBInfinityLocalManager.m
//  YouboraLib
//
//  Created by Enrique Alfonso Burillo on 18/07/2018.
//  Copyright © 2018 NPAW. All rights reserved.
//

#import "YBInfinityLocalManager.h"

#import "YBChrono.h"

//Key names
NSString *const PREFERENCES_SESSION_ID_KEY = @"session_id";
NSString *const PREFERENCES_CONTEXT_KEY = @"context_id";
NSString *const PREFERENCES_LAST_ACTIVE_KEY = @"last_active_id";

@interface YBInfinityLocalManager()

@property(nonatomic, strong) NSUserDefaults *userPrefs;

@end

@implementation YBInfinityLocalManager

- (instancetype) init {
    self = [super init];
    if (self) {
        self.userPrefs = [NSUserDefaults standardUserDefaults];
    }
    return self;
}

- (void) saveSessionWithId: (NSString *) sessionId {
    [self saveValueWithKey:PREFERENCES_SESSION_ID_KEY andValue: sessionId];
}

- (NSString *) getSessionId {
    return [self getValueWithKey:PREFERENCES_SESSION_ID_KEY];
}

- (void) saveContextWithContext: (NSString *) context {
    [self saveValueWithKey:PREFERENCES_CONTEXT_KEY andValue:context];
}

- (NSString *) getContext {
    return [self getValueWithKey:PREFERENCES_CONTEXT_KEY];
}

- (void) saveLastActiveDate {
    [self saveNumberWithKey:PREFERENCES_LAST_ACTIVE_KEY andValue:@([YBChrono getNow])];
}

- (NSNumber *) getLastActive {
    return [self getNumberWithKey:PREFERENCES_LAST_ACTIVE_KEY];
}

- (void) saveValueWithKey: (NSString *) key andValue: (NSString *) value{
    [self.userPrefs setValue:value forKey:key];
}

- (NSString *) getValueWithKey: (NSString *) key {
    return [self.userPrefs objectForKey:key];
}

- (void) saveNumberWithKey: (NSString *) key andValue: (NSNumber *) value {
    [self.userPrefs setObject:value forKey:key];
}

- (NSNumber *) getNumberWithKey: (NSString *) key {
    return [self.userPrefs objectForKey:key];
}

@end

File: YouboraLib/persistence/infinity/YBInfinityLocalManager.swift
//
//  YBInfinityLocalManager.swift
//  YouboraLib iOS
//
//  Created by Tiago Pereira on 04/05/2020.
//  Copyright © 2020 NPAW. All rights reserved.
//

import Foundation

@objcMembers public class YBInfinityLocalManager: NSObject {
    
    /**
    * Saves new session id on <NSUserDefaults>
    * @param sessionId sessionId to save
    */
    public static func saveSession(sessionId: String) {
        UserDefaults.standard.set(sessionId, forKey: YBConstants.preferencesSessionIdKey)
    }
    
    /**
    * Gets saved session id on <NSUserDefaults>
    * @return session id to save
    */
    public static func getSessionId() -> String? {
        return UserDefaults.standard.value(forKey: YBConstants.preferencesSessionIdKey) as? String
    }
    
    /**
    * Saves new context on <NSUserDefaults>
    * @param context to save
    */
    public static func saveContext(context: String) {
        UserDefaults.standard.set(context, forKey: YBConstants.preferencesContextKey)
    }
    
    /**
    * Gets saved context on <NSUserDefaults>
    * @return context to save
    */
    public static func getContext() -> String? {
        return UserDefaults.standard.value(forKey: YBConstants.preferencesContextKey) as? String
    }
    
    /**
    * Saves timestamp of last event sent on <NSUserDefaults>
    */
    public static func saveLastActiveDate() {
        UserDefaults.standard.set(NSNumber(value: YBChrono().now), forKey: YBConstants.preferencesLastActiveKey)
    }
    
    /**
    * Gets saved timestamp on <NSUserDefaults>
    * @return context to save
    */
    public static func getLastActive() -> NSNumber? {
        return UserDefaults.standard.value(forKey: YBConstants.preferencesLastActiveKey) as? NSNumber
    }
    
    /**
    * Method that will clean all data in the user defaults realtive with
    * infinity local manager
    */
    public static func cleanLocalManager() {
        DispatchQueue.main.async {
            UserDefaults.standard.removeObject(forKey: YBConstants.preferencesSessionIdKey)
            UserDefaults.standard.removeObject(forKey: YBConstants.preferencesContextKey)
            UserDefaults.standard.removeObject(forKey: YBConstants.preferencesLastActiveKey)
            UserDefaults.standard.synchronize()
        }
    }
}
